subsys/usb/class/bt_h4.c

```diff
@@ -0,0 +1,325 @@
+#include <usb_descriptor.h>
+
+#include <net/buf.h>
+
+#include <bluetooth/buf.h>
+#include <bluetooth/hci_raw.h>
+#include <bluetooth/l2cap.h>
+
+#define LOG_LEVEL CONFIG_USB_DEVICE_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(usb_bt_h4);
+
+static K_FIFO_DEFINE(rx_queue);
+static K_FIFO_DEFINE(tx_queue);
+
+/* ACL data TX buffers */
+#if defined(CONFIG_BT_CTLR_TX_BUFFERS)
+#define BUF_COUNT MAX(CONFIG_BT_HCI_CMD_COUNT, CONFIG_BT_CTLR_TX_BUFFERS)
+#else
+#define BUF_COUNT MAX(CONFIG_BT_HCI_CMD_COUNT, 4)
+#endif
+
+#if defined(CONFIG_BT_CTLR_TX_BUFFER_SIZE)
+#define BT_L2CAP_MTU (CONFIG_BT_CTLR_TX_BUFFER_SIZE - BT_L2CAP_HDR_SIZE)
+#else
+#define BT_L2CAP_MTU 64
+#endif
+
+/* Data size needed for CMD and ACL buffers */
+#define BUF_SIZE MAX(BT_BUF_RX_SIZE, BT_L2CAP_BUF_SIZE(BT_L2CAP_MTU))
+NET_BUF_POOL_DEFINE(tx_pool, BUF_COUNT, BUF_SIZE, sizeof(u8_t), NULL);
+
+#define BT_H4_OUT_EP_ADDR		0x01
+#define BT_H4_IN_EP_ADDR		0x81
+
+#define BT_H4_OUT_EP_IDX		0
+#define BT_H4_IN_EP_IDX			1
+
+/* HCI RX/TX threads */
+static K_THREAD_STACK_DEFINE(rx_thread_stack, 512);
+static struct k_thread rx_thread_data;
+static K_THREAD_STACK_DEFINE(tx_thread_stack, 512);
+static struct k_thread tx_thread_data;
+
+struct usb_bt_h4_config {
+	struct usb_if_descriptor if0;
+	struct usb_ep_descriptor if0_out_ep;
+	struct usb_ep_descriptor if0_in_ep;
+} __packed;
+
+USBD_CLASS_DESCR_DEFINE(primary, 0) struct usb_bt_h4_config bt_h4_cfg = {
+	/* Interface descriptor 0 */
+	.if0 = {
+		.bLength = sizeof(struct usb_if_descriptor),
+		.bDescriptorType = USB_INTERFACE_DESC,
+		.bInterfaceNumber = 0,
+		.bAlternateSetting = 0,
+		.bNumEndpoints = 2,
+		.bInterfaceClass = CUSTOM_CLASS, /* TBD */
+		.bInterfaceSubClass = 0,
+		.bInterfaceProtocol = 0,
+		.iInterface = 0,
+	},
+
+	/* Data Endpoint OUT */
+	.if0_out_ep = {
+		.bLength = sizeof(struct usb_ep_descriptor),
+		.bDescriptorType = USB_ENDPOINT_DESC,
+		.bEndpointAddress = BT_H4_OUT_EP_ADDR,
+		.bmAttributes = USB_DC_EP_BULK,
+		.wMaxPacketSize = sys_cpu_to_le16(CONFIG_BT_H4_BULK_EP_MPS),
+		.bInterval = 0x01,
+	},
+
+	/* Data Endpoint IN */
+	.if0_in_ep = {
+		.bLength = sizeof(struct usb_ep_descriptor),
+		.bDescriptorType = USB_ENDPOINT_DESC,
+		.bEndpointAddress = BT_H4_IN_EP_ADDR,
+		.bmAttributes = USB_DC_EP_BULK,
+		.wMaxPacketSize = sys_cpu_to_le16(CONFIG_BT_H4_BULK_EP_MPS),
+		.bInterval = 0x01,
+	},
+};
+
+#define H4_CMD 0x01
+#define H4_ACL 0x02
+#define H4_SCO 0x03
+#define H4_EVT 0x04
+
+static struct usb_ep_cfg_data bt_h4_ep_data[] = {
+	{
+		.ep_cb = usb_transfer_ep_callback,
+		.ep_addr = BT_H4_OUT_EP_ADDR,
+	},
+	{
+		.ep_cb = usb_transfer_ep_callback,
+		.ep_addr = BT_H4_IN_EP_ADDR,
+	},
+};
+
+static void bt_h4_read(u8_t ep, int size, void *priv)
+{
+	struct net_buf *buf = priv;
+
+	if (size > 0) {
+		buf->len += size;
+		net_buf_put(&tx_queue, buf);
+		buf = NULL;
+	}
+
+	if (buf) {
+		net_buf_unref(buf);
+	}
+
+	buf = net_buf_alloc(&tx_pool, K_FOREVER);
+	__ASSERT_NO_MSG(buf);
```

Comment:
With ASSERT disabled that will be never caught, also not in net_buf_simple_reserve(). I know there is the same code in acl_read_cb, but why not check explicitly for NULL and log the error?

Reply:
Or use BT_ASSERT, which are always enabled.

Reply:
Right, either this is no expected behavior since we are using k_FOREVER but I can use BT_ASSERT instead.

Reply:
Looks like bringing BT_ASSERT would possible require making it part of a public API so I will leave this for now and just enable asserts on the sample.

Reply:
I do not think it is acceptable from security perspective, either it has to be checked against NULL (`if (buf == NULL)`) or not (`__ASSERT_NO_MSG(buf);` not necessary at all) regardless of the build options/config.

Reply:
Alright then I would removed since using K_FOREVER should never really fail here.